Improved REPL

finnbar opened the pull request:
REPL has been improved to make it more useful. Changes are as follows:

* Added file loading via `:c`, which previously did nothing. (The `compile` function returned unit before.)
* Allow definitions to be written on the REPL for later use, via `let x = <expr>` just like in an actual program. Expressions are also allowed, and are still prefaced by an implicit `let it =`.

finnbar committed:
Update details in package.yaml for PR

finnbar committed:
Make `compile` actually import a file

finnbar committed:
Make `compile` actually import a file

finnbar committed:
Fixed deprecated imports

finnbar committed:
REPL can now parse definitions

finnbar committed:
Added maintainer email

finnbar committed:
Remove empty README heading

mbg reviewed (commented):
LGTM, but probably will need rebasing after #2 is merged as well.

finnbar committed:
Removed accidental files from stack-work

finnbar committed:
Fixed details in package.yaml

finnbar commented:
Just realised that this branch has an old Parser.hs on it, since I didn't regenerate it with happy after changing the parser. Will delete the Lexer.hs and Parser.hs files after #2 is merged, which solves this problem.

finnbar committed:
Fixed me being bad at rebase

finnbar committed:
Removed Lexer.hs and Parser.hs

finnbar committed:
Merge branch 'master' into better-repl

finnbar commented:
Okay, this has been brought up to date with master despite my best attempts - so should be ready to merge.

mbg commented on src/Interpreter.lhs:
```diff
@@ -129,7 +130,13 @@
     
 >   eval :: GlEnv -> Expr -> Either String Expr
->   eval env e = evalState (runErrorT (eval' e)) env
+>   eval env e = evalState (runExceptT (eval' e)) env
+
+>   evalDef :: GlEnv -> Definition -> Either String GlEnv
+>   evalDef env (Def var e) = do
+>       ty <- infer e (toTyEnv env)
+>       let gldef = GlDef {glExpr = e, glType = ty}
+>       return $ (var, gldef) : env
```
I am wondering if there is a better name for this function than `evalDef`, given that it doesn't seem to be doing any evaluation. How about something like `addDef`? Indeed, that already exists in https://github.com/mbg/simply-typed-ext/blob/3b91d874260a87169a43265489e17af3ba532e67/app/Main.lhs#L35

finnbar committed:
evalDef superceded by addDef
They technically throw errors in different ways, but addDef is how it is
done elsewhere.
